### PR TITLE
TASK-1004 ignore large inputs

### PR DESCRIPTION
**Pull request by Tianhao-Gu**

### CODE REVIEW

File: test/kb_quast_server_test.py [hunk 1]
```diff
@@ -216,6 +281,7 @@ def test_fail_bad_ws_ref(self):
                 self.ws_info[0], self.ws_info[1]),
             exception=WorkspaceError)
 
+    @unittest.skip("skipped test_fail_bad_ws_type")
```
Comment (Tianhao-Gu) [inline test/kb_quast_server_test.py]: I think this test is no longer valid.

**Comment by Tianhao-Gu** [comment]
@MrCreosote Hi Gavin, this is PR for https://kbase-jira.atlassian.net/browse/TASK-1004. 

Thanks!

File: test/kb_quast_server_test.py (hunk 1 above)
Comment (MrCreosote) [inline test/kb_quast_server_test.py]: Could you explain why real quick?

**Comment by Tianhao-Gu** [comment]
@MrCreosote oh yea yea. Sry for not understanding the task >_<

File: lib/kb_quast/kb_quastImpl.py [hunk 2]
```diff
@@ -128,14 +133,24 @@ def run_quast_exec(self, outdir, filepaths, labels):
         # quast will ignore bad files and keep going, which is a disaster for
         # reproducibility and accuracy if you're not watching the logs like a hawk.
         # for now use this hack to check that all files were processed. Maybe there's a better way.
-        files_proc = len(_os.listdir(_os.path.join(outdir, 'predicted_genes'))) / 2
+        files_proc = len(open(_os.path.join(outdir, 'report.tsv'), 'r').readline().split('\t')) - 1
```
Comment (Tianhao-Gu) [inline lib/kb_quast/kb_quastImpl.py]: with no --glimmer, it doesn't produce predicted_genes folder.

File: test/kb_quast_server_test.py [hunk 3]
```diff
@@ -226,12 +254,11 @@ def test_fail_bad_ws_type(self):
                                             'objects':
                                             [bad_object_type]})[0]
         bo_type = bad_object[2]
-        md5type = self.ws.translate_to_MD5_types([bo_type])[bo_type]
         bad_object_ref = str(bad_object[6]) + '/' + str(bad_object[0]) + '/' + str(bad_object[4])
         self.fail_quast(
             {'assemblies': [bad_object_ref]},
-            "Invalid type! Expected one of ['KBaseGenomes.ContigSet', " +
-            "'KBaseGenomeAnnotations.Assembly'], received {}".format(md5type),
+            "Cannot write data to fasta; invalid WS type ({}).  ".format(bo_type) +
```
Comment (Tianhao-Gu) [inline test/kb_quast_server_test.py]: Seems like, AssemblyUtil changes exception message.

**Comment by Tianhao-Gu** [comment]
@MrCreosote Could you take a look at this PR again when you have a minute? thanks

File: lib/kb_quast/kb_quastImpl.py [hunk 4]
```diff
@@ -274,9 +291,11 @@ def run_QUAST(self, ctx, params):
                 filepaths.append(p)
                 labels.append(l)
 
+        skip_glimmer = self.check_large_input(filepaths)
+
```
Comment (MrCreosote) [inline lib/kb_quast/kb_quastImpl.py]: Should log that you're skipping glimmer either here or in the exec or check_large methods

File: lib/kb_quast/kb_quastImpl.py [hunk 5]
```diff
@@ -128,14 +134,25 @@ def run_quast_exec(self, outdir, filepaths, labels):
...
+    def check_large_input(self, filepaths):
+        skip_glimmer = False
+
+        for filepath in filepaths:
+            records = list(_SeqIO.parse(filepath, 'fasta'))
+            if len(records) > self.TEN_MB:
+                skip_glimmer = True
+                break
+
+        return skip_glimmer
```
Comment (MrCreosote) [inline lib/kb_quast/kb_quastImpl.py]: Is it decided that if any of the assemblies are > 10Mb glimmer should be disabled? I don't see an answer to your question in the JIRA ticket.

File: test/kb_quast_server_test.py [hunk 6]
```diff
@@ -267,6 +293,21 @@ def test_quast_app(self):
                                                  'workspace_name': self.ws_info[1]})[0]
         self.check_quast_app_output(ret, 315170, 315200, '6aae4f232d4d011210eca1965093c22d',
                                     '2010dc270160ee661d76dad6051cda32')
+    
+    @patch.object(kb_quast, "check_large_input", return_value=True)
```
Comment (MrCreosote) [inline test/kb_quast_server_test.py]: Unless I'm missing something check_large_input is always mocked and so never actually tested that a large input will result in a return of True. IMO at least one test should exercise the method for each of the two possible outputs.

**Comment by MrCreosote** [comment]
Also looks like the build is borked - someone must've checked in a travis file that doesn't work.

**Comment by dcchivian** [comment]
I am fine with using the 10 megabyte threshold instead of 10 megabases (if for some reason the latter is not available easily).

**Comment by dcchivian** [comment]
@Tianhao-Gu the Travis CI error is the rakefile (ruby dependency) that goes away when you use the right travis.yml.  Just steal one from anywhere (e.g. https://github.com/kbaseapps/kb_SetUtilities/blob/master/.travis.yml )

**Comment by Tianhao-Gu** [comment]
@dcchivian I can find base count with Biopython library pretty easily. Are you okay if any file is larger than 10 megabases, we skip --glimmer for all files? 

--glimmer is a binary option. We can either turn it on for all files. Or turn it off for all files.

**Comment by dcchivian** [comment]
@Tianhao-Gu @scanon You could check filesize > 15 Mbytes first so you don't have to bother with base counting for obviously huge files.  I would make gene calls OFF the default behavior above threshold   and ON the default behavior below threshold.  If you really want, you can add an option to allow force override for large files gene calls ON (and have that forced override be OFF by default).

**Comment by Tianhao-Gu** [comment]
@MrCreosote Could you take another look? Many discussions happened here: https://kbase-jira.atlassian.net/browse/PTV-465

To summarize, added a 'force_glimmer' option to run glimmer regardless of file size. If force_glimmer is False, when any file has more than 10MBases, we skip glimmer on all files.

File: lib/kb_quast/kb_quastImpl.py [hunk 7]
```diff
@@ -128,14 +135,25 @@ def run_quast_exec(self, outdir, filepaths, labels):
...
             err = ('QUAST skipped some files - {} expected, {} processed.'
                    .format(files_exp, files_proc))
             self.log(err)
             raise ValueError(err)
 
+    def check_large_input(self, filepaths):
+        skip_glimmer = False
+
+        for filepath in filepaths:
+            basecount = len(list(_SeqIO.parse(filepath, 'fasta')))
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply lib/kb_quast/kb_quastImpl.py]: SeqIO.parse returns a list of dicts representing each contigs. The len is the count of contigs. But looks like I misunderstood the issue. I will do the sum logic.

**Comment by Tianhao-Gu** [comment]
@MrCreosote Updated thresholds to 20MB. and counting all length of contigs for all files.

File: ui/narrative/methods/run_QUAST_app/display.yaml [hunk 8]
```diff
@@ -24,6 +24,10 @@ parameters :
         ui-name : Assemblies
         short-hint : Assembly objects to be analyzed by QUAST.
 
+    force_glimmer:
+        ui-name : Force Gene Calls For Large Assemblies
```
Comment (MrCreosote) [inline ui/narrative/methods/run_QUAST_app/display.yaml]: It's not just large assemblies, it's if the total number of bases across all assemblies is large - not sure if this is accurate

Comment (Tianhao-Gu) [reply ui/narrative/methods/run_QUAST_app/display.yaml]: @dcchivian Should we just put "Force Gene Calls"?

**Comment by dcchivian** [comment]
The problem with "force gene calls" is that it doesn't sufficiently clue the user into the danger.  Stating that it's for large assemblies makes them think a bit more and increases their understanding in an at-a-glance way (i.e. without resorting to looking at documentation, which 90% of people won't do).  In this case the danger is that they will select it, costing potentially days of runtime, so the cost/benefit analysis indicates we should REALLY try to prevent them from selecting it unless they really want it and are aware of the time implications.

There is a subtlety that if there are multiple small assemblies, then the word "large assemblies" is misleading.  I only cared about not running GlimmerHMM for individual large MG assemblies, and didn't have an opinion on many small assemblies (say 10-100 isolate bacteria).  This seems to be the case where the wording would be slightly off.  I can't come up with subtler wording that wouldn't really start being too long, so how about we include the distinction in the "hint"?  Anyway, the take home is use: "force gene calls for large assemblies" and say something about it also applies to many small assemblies in the hint.

**Comment by Tianhao-Gu** [comment]
@dcchivian @MrCreosote I put some explanation on hint. Wondering if you guys are good with this.

**Comment by MrCreosote** [comment]
Good enough for me.

**Pull request merged**